lamagraph-compiler/src/LamaML/Syntax.hs

```diff
@@ -0,0 +1,290 @@
+{- |
+
+LamaML syntax
+-}
+module LamaML.Syntax (
+    -- * Grammar rules
+    -- $rules
+
+    -- * Lexical conventions
+
+    -- ** Blanks
+    -- $lexing_blanks
+
+    -- ** Comments
+    -- $lexing_comments
+
+    -- ** Identifiers
+    -- $lexing_idents
+
+    -- ** Integer literals
+    -- $lexing_int_lits
+
+    -- ** Character literals
+    -- $lexing_char_lits
+
+    -- ** String literals
+    -- $lexing_string_lits
+
+    -- ** Operators
+    -- $lexing_operators
+
+    -- ** Keywords
+    -- $lexing_keywords
+
+    -- * Names
+    -- $names
+
+    -- * Type expressions
+    -- $types
+
+    -- * Constants
+    -- $constants
+
+    -- * Patterns
+    -- $patterns
+
+    -- * Expressions
+    -- $expressions
+
+    -- * Type definitions
+    -- $typedefs
+
+    -- * Declarations
+    -- $decls
+) where
+
+{- $rules
+Grammar rules are written in monospace font.
+Italicized words represent @/nonterminals/@.
+@Terminals@ are written in upright font.
+@( )@ mean grouping, @{ }@ mean zero or more repetitions, @[ ]@ mean optional.
+In case of clash between grammar and terminal meaning, terminal symbol will be typeset as @__bold monospace__@.
+-}
+
+{- $lexing_blanks
+Spaces, tabs, @\\n@ and @\\r@ are ignored and are used to separate tokens.
+-}
+
+{- $lexing_comments
+Comments must be enclosed in @(*@ and @*)@.
+Nesting is allowed and must be handled.
+-}
+
+{- $lexing_idents
+@
+/ident/ ::= ( /letter/ | _ ) { /letter/ | 0...9 | _ | ' }
+
+/capitalized-ident/ ::= ( A...Z ) { /letter/ | 0...9 | _ | ' }
+
+/lowercase-ident/ ::= ( a...z | _ ) { /letter/ | 0...9 | _ | ' }
+
+/letter/ ::= A...Z | a...z
+@
+-}
+
+{- $lexing_int_lits
+@
+/integer-literal/ ::= \[-] ( 0...9 ) { 0...9 | _ }
+@
+
+/Do we need hex\/octal\/binary numbers and\/or size suffixes?/
```

Comment:
I guess, that size suffix can be useful. Hex, binary can be omitted.

Reply:
What sizes do we want to support? Standard 32 and 64 bit? Or do we want even 8 and 16 bit wide integers? Also what about signed/unsigned suffixes?

Reply:
I think that 32 and 64 bit, signed and unsigned. At the first stage. To realize problems with values of different sizes.